### ISSUE TITLE
Show GitHub metrics within pub.dev

### ISSUE THREAD

**Issue by jayoung-lee** [dart-lang/pub-dev#4369]
When evaluating a package that needs regular maintenance, users visit GitHub to check the package’s recent activities and maintenance status. We can consider showing key GitHub metrics within pub.dev to save users’ trip to GitHub.

Key metrics that users mentioned in a recent user study were:
* Last commit date/content
* Number of open issues 
* Number of open PRs

<img width="381" alt="sample4" src="https://user-images.githubusercontent.com/11873270/102841341-9bfd1f80-43b9-11eb-9254-cee53dc30ab2.png">

**Comment by YazeedAlKhalaf** [comment]
That's a really great and useful idea!
I believe we could use the GitHub API to do this. `npm` already does this and even VSCode marketplace.

**Comment by elbeicktalat** [comment]
Very nice idea in this way developers safe time, also having this data will make easier choices between the similar package.

**Comment by stuartmorgan** [comment]
One complicating factor here is packages from monorepos (flutter/plugins and flutter/packages being prominent examples, but certainly not the only ones); repo-level metrics will be misleading for per-plugin evaluation.

**Comment by jayoung-lee** [comment]
@stuartmorgan  That's a good point. I think we shouldn't display the metrics when there's no 1:1 match. Do you know how many packages are from monorepos?

**Comment by stuartmorgan** [comment]
I don't, although we could probably cobble together a script to make an educated guess based on crawling pub.dev links. The problem with not showing anything would be that it would likely be confusing to have a lot of the most popular packages (from a quick manual check, it looks like about 50% of Flutter Favorites, for instance) not have the quality metrics.

**Comment by leighajarett** [comment]
+1 This came up during research sessions for Flutter Vikings. Developers were also interested in seeing security vulnerabilities broken out as their own category if possible (perhaps through the use of Github tags). 

We also talked to developers about being too intertwined with Github, but everyone we spoke with was happy to only use Github for hosting their packages as opposed to other git providers. 

@stuartmorgan in the case of monorepos: could we also try to look only for commits in directories that match the package name, and issue tags that match as well (or contain the name so it would catch our p: <package-name>)? I know there would be some errors here, but maybe we just make that clear to both package consumers and developers so they're aware

**Comment by stuartmorgan** [comment]
> could we also try to look only for commits in directories that match the package name

We'd want to spot-check some of the other monorepos, but that could likely work. (There are some subtle questions to resolve first about what exactly we even *want* to count when it comes to federated plugins.)

> issue tags that match as well

Maybe? I just checked the Plus Plugins repo, and they don't appear to be labeling consistently.

I actually have significant reservations about issue counts as a metric we actively promote though. In practice, issue counts are measures of five things:
1. popularity - Flutter has 11k open issues not because it's orders of magnitude less well-maintained than most packages, but because orders of magnitude more people are using it and filing issues. Obviously that's an extreme example, but the principle applies within the package ecosystem.
1. size/scope - Same argument here.
1. age - (maybe; see next item) Flutter has more than 2x the issues it had when I started working on it; that's not because it's become more poorly maintained, we just get more reports than we can fix (which is common, and not inherently problematic) so accumulation is a function of time.
1. issue handling policy - I'm aware of some packages that run a bot in their repo that closes every bug that hasn't been updated in several weeks. That doesn't make them better maintained than packages with a policy of leaving all unresolved bugs or feature requests open (like Flutter mostly does), but changes issue counts dramatically.
1. bug reports and feature requests - there's no reliable way to distinguish between them.

So of five things that have *major* influence on issue count, only one (really, only part of one) is really any kind of indication of package quality/maintenance quality. There's so much noise that at best I think the number is meaningless. At worst, we would incentive behavior that I would argue is undesirable, like running bots that close almost everything (which results in things like 20 one-off closed issues about the same thing, rather than one issue that collects discussion, votes, etc.)

**Comment by sigurdm** [comment]
I agree it is hard to tell anything for certain from many of these numbers, and many of them can be gamed to some extent.

Still many people use the github link to estimate how well-maintained a package is. So we should be able to extract *something* useful if we're careful.

* Maybe **rate** of commits (how many commits per week during the last year) could be a useful indication of project liveliness. github has an api for getting recent commit activity,
* Github stars: indicating how many people are interested.

**Comment by stuartmorgan** [comment]
If we are willing to do not-trivial queries, an interesting metric could be something based on the percentage of recent issue that a project member has interacted with in some way (comment, label, etc.) That would give some indication of whether someone is at least looking at issues.

**Comment by stuartmorgan-g** [comment]
> * Package scores should **decline/decay** over time without updates

This would be actively harmful for some packages. For instance, Dart math libraries can be fine for a very long time without updates, for examples, because they often don't have dependencies and math itself doesn't change. Similarly, `mustache_template` hasn't been updated in 4 years, and it's still actively used in the current version of the `flutter` tool with no issues.

It is easy to describe general heuristics that would apply well to specific examples, but don't generalize well.

**Comment by mhamri** [comment]
> `mustache_template`

That’s a very bad example for your argument but perfect example for mine. https://github.com/jonahwilliams/mustache is just a fork of the original repository (which also isn’t maintained). Forks don’t have an issue tab, so you literally have no idea how many bugs are hanging around. It’s not maintained, and it has six PRs that have never been merged since 2020. The original repo itself has 16 unresolved issues that still haven’t been addressed.

Saying “we can use it” or that it should only score five points lower than mustachex are two different things. The score should decay over time. When searching for similar solutions, if there’s only one option, people will use it. But if there are multiple options and one is clearly better maintained, why would we ever choose the one that isn’t? Why should they ever get a similar score?

If something like Dart 3.0 compatibility becomes an issue, which carries the higher risk: a package that hasn’t been updated in four years, or a package doing the same job but actively maintained? Anyone with even a little sense of risk wouldn’t choose an unmaintained package, no matter how well it worked in the past. It’s risky, and nothing can make it less risky—it can break everything, including every package that depends on it. Pub.dev shouldn’t promote these packages as the best possible solution for that need; they should be at the bottom of the list.

Also, packages like dart:math are part of the framework itself. They’re published with each version of Dart, so I don’t think those will be affected. If they were, then exceptions could be considered.

**Comment by stuartmorgan-g** [comment]
> Also, just because someone from the Flutter team maintains a package doesn’t automatically mean it’s well maintained. For example, that particular repository doesn’t even have an Issues tab.

None of that has anything to do with penalizing package scores for not having published recently, which is the specific suggestion my comment was replying to.

**Comment by stuartmorgan-g** [comment]
> If you want to rank packages with similar functionality by giving higher scores to those with better maintenance, your calculation will naturally assign lower scores to unmaintained packages. That, in effect, is a form of decay.

- Pana scores and pub.dev search rank are not the same thing.
- pub.dev does not know if two packages provide "similar functionality".
- pub.dev does not know if packages are unmaintained, and it is *extremely hard* to define objective criteria which correctly capture whether a package is, conceptually, "maintained", which is my central point.

**Comment by isoos** [comment]
> We’re really talking about how packages should be sorted and ranked, and the current method seems flawed. From my experience working with many packages, I’ve noticed countless forks of old ones floating around on pub.dev, often in questionable states.

Please file new issues describing the search queries and what you find bad in the results. We will look into them and see if/how we can improve it.

> I’ve even submitted PRs to maintainers who never reviewed or merged them, forcing me to either find yet another package or publish my own fork. This cycle is frustrating. Constantly switching packages and digging through inconsistent documentation is exhausting, and it significantly hurts the developer experience

I see that this can be frustrating, but life can happen to anybody, and sometimes your request may coincide with a busy period. Neither the past good, nor the past bad periods will be representative on the short term outcome, though, of course there is some correlation there. I think it is worth to expose some of it as metrics, but we don't see enough evidence that this should be a significant part of the ranking. By proxy, the package's use (likes and downloads) will probably also include it, since a badly maintained package won't get as much use as a better one. Not in the long term anyway.
